subsys/bluetooth/host/classic/avrcp.c

```diff
@@ -0,0 +1,460 @@
+	return 0;
+}
+
+struct bt_avrcp *bt_avrcp_connect(struct bt_conn *conn)
+{
+	struct bt_avrcp *avrcp;
+	int err;
+
+	avrcp = get_new_connection(conn);
+	if (!avrcp) {
+		LOG_ERR("Cannot allocate memory");
+		return NULL;
+	}
+
+	avrcp->session.ops = &avctp_ops;
+	err = bt_avctp_connect(conn, &(avrcp->session));
+	if (err < 0) {
+		/* If error occurs, undo the saving and return the error */
+		memset(avrcp, 0, sizeof(struct bt_avrcp));
+		LOG_DBG("AVCTP Connect failed");
+		return NULL;
+	}
+
+	LOG_DBG("Connection request sent");
+	return avrcp;
+}
+
+int bt_avrcp_disconnect(struct bt_avrcp *avrcp)
+{
+	int err;
+
+	err = bt_avctp_disconnect(&(avrcp->session));
+	if (err < 0) {
+		LOG_DBG("AVCTP Disconnect failed");
+		return err;
+	}
+
+	return err;
+}
+
+static struct net_buf *avrcp_create_pdu(struct bt_avrcp *avrcp, bt_avctp_cr_t cr)
+{
+	struct net_buf *buf;
+
+	buf = bt_avctp_create_pdu(&(avrcp->session), cr, BT_AVCTP_PKT_TYPE_SINGLE,
+				  BT_AVCTP_IPID_NONE, &avrcp->local_tid,
+				  sys_cpu_to_be16(BT_SDP_AV_REMOTE_SVCLASS));
+
+	return buf;
+}
+
+static struct net_buf *avrcp_create_unit_pdu(struct bt_avrcp *avrcp, bt_avctp_cr_t cr)
+{
+	struct net_buf *buf;
+	struct bt_avrcp_unit_info_cmd *cmd;
+
+	buf = avrcp_create_pdu(avrcp, cr);
+	if (!buf) {
+		return buf;
+	}
+
+	cmd = net_buf_add(buf, sizeof(*cmd));
+	memset(cmd, 0, sizeof(*cmd));
+	BT_AVRCP_HDR_SET_CTYPE(&cmd->hdr, cr == BT_AVCTP_CMD ? BT_AVRCP_CTYPE_STATUS
+							     : BT_AVRCP_CTYPE_IMPLEMENTED_STABLE);
+	BT_AVRCP_HDR_SET_SUBUNIT_ID(&cmd->hdr, BT_AVRCP_SUBUNIT_ID_IGNORE);
+	BT_AVRCP_HDR_SET_SUBUNIT_TYPE(&cmd->hdr, BT_AVRCP_SUBUNIT_TYPE_UNIT);
+	cmd->hdr.opcode = BT_AVRCP_OPC_UNIT_INFO;
+
+	return buf;
+}
+
+static int avrcp_send(struct bt_avrcp *avrcp, struct net_buf *buf)
+{
+	int err;
+	struct bt_avctp_header *avctp_hdr = (struct bt_avctp_header *)(buf->data);
+	struct bt_avrcp_header *avrcp_hdr =
+		(struct bt_avrcp_header *)(buf->data + sizeof(*avctp_hdr));
+	uint8_t tid = BT_AVCTP_HDR_GET_TRANSACTION_LABLE(avctp_hdr);
+	bt_avctp_cr_t cr = BT_AVCTP_HDR_GET_CR(avctp_hdr);
+	bt_avrcp_ctype_t ctype = BT_AVRCP_HDR_GET_CTYPE(avrcp_hdr);
+	bt_avrcp_subunit_type_t subunit_id = BT_AVRCP_HDR_GET_SUBUNIT_ID(avrcp_hdr);
+
+	LOG_DBG("AVRCP send cr:0x%X, tid:0x%X, ctype: 0x%X, opc:0x%02X\n", cr, tid, ctype,
+		avrcp_hdr->opcode);
+	err = bt_avctp_send(&(avrcp->session), buf);
+	if (err < 0) {
+		return err;
+	}
+
+	if (cr == BT_AVCTP_CMD && avrcp_hdr->opcode != BT_AVRCP_OPC_PASS_THROUGH) {
+		avrcp->req.tid = tid;
+		avrcp->req.subunit = subunit_id;
+		avrcp->req.opcode = avrcp_hdr->opcode;
+
+		k_work_reschedule(&avrcp->timeout_work, AVRCP_TIMEOUT);
+	}
+
+	return 0;
+}
+
+int bt_avrcp_get_unit_info(struct bt_avrcp *avrcp)
+{
+	struct net_buf *buf;
+	uint8_t param[5];
+
+	buf = avrcp_create_unit_pdu(avrcp, BT_AVCTP_CMD);
+	if (!buf) {
+		return -ENOMEM;
+	}
+
+	memset(param, 0xFF, 5);
+	net_buf_add_mem(buf, param, sizeof(param));
+
+	return avrcp_send(avrcp, buf);
+}
```

Comment:
Is there something missing in the functionality of this function? I don't see anywhere handling the response of this command. From Api perspective, the result should be notified if the function is called from application.

Reply:
Not really now, since the unit info is actually fixed in AVRCP.
As said in AVRCP v1.6.2, the TG device shall return a unit_type PANEL(0x09) in the response frame.
In the mojority cases, we do not care about this return value, or we can add a error handling if the response is incorrect.

Reply:
Done.